### PR TITLE
IGNITE-21797: Column's type mismatch during marshalling of pojo (embedded API)

### PR DESCRIPTION
**Pull request by lowka**
- Updates Marshaller API to support random access for reading/writing fields and extraction of key / value objects, because key/value columns and be placed anywhere inside a row and neither key nor values columns constitute continuous blocks.
- Updates BinaryTupleSchema API to support aforementioned configuration.

https://issues.apache.org/jira/browse/IGNITE-21797

---

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerReader.java [hunk 1]
```diff
@@ -30,6 +30,18 @@
  * Binary reader.
  */
 public interface MarshallerReader {
+    /**
+     * Set reader's position to the given value.
+     */
+    void setIndex(int index);
```
Comment (korlov42) [inline modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerReader.java:36]: if we want to extend MarshallerReader with ability to read arbitrary field, it's better to extend every `read` method with field index param rather that change its internal state every time.

File: modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/Marshaller.java [hunk 2]
```diff
@@ -267,13 +306,36 @@ public Object readObject(MarshallerReader reader, Object target) throws Marshall
             return obj;
         }
 
+        /** {@inheritDoc} */
+        @Override
+        public Object readValue(MarshallerReader reader, Object target) throws MarshallerException {
```
Comment (korlov42) [inline modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/Marshaller.java:311]: not sure I fully understand the difference between `readObject` and `readValue`...

File: modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerColumn.java [hunk 3]
```diff
@@ -68,12 +70,32 @@ public MarshallerColumn(String name, BinaryMode type) {
...
+        this.defValSup = defValSup == null ? NULL_SUPPLIER : defValSup;
+        this.scale = scale;
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param name      Column name.
+     * @param type      An instance of column data type.
+     * @param defValSup Default value supplier.
```
[the remark this answers is not in the document]
Comment (lowka) [reply modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerColumn.java:77]: Thanks fixed.

File: modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerColumn.java [hunk 4]
```diff
@@ -68,12 +70,32 @@ public MarshallerColumn(String name, BinaryMode type) {
      * @param defValSup Default value supplier.
      */
     public MarshallerColumn(String name, BinaryMode type, @Nullable Supplier<Object> defValSup, int scale) {
```
[the remark this answers is not in the document]
Comment (lowka) [reply modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerColumn.java]: Remove this CTOR, marked another constructor (which does not provide all the fields) as TestOnly.

File: modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerColumn.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerColumn.java:77]: nitpicking: and what about `scale`? :)

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/KvMarshallerImpl.java [hunk 5]
```diff
@@ -118,7 +129,7 @@ public K unmarshalKey(Row row) throws MarshallerException {
     @Nullable
     @Override
     public V unmarshalValue(Row row) throws MarshallerException {
-        Object o = valMarsh.readObject(new RowReader(row, schema.keyColumns().size()), null);
+        Object o = valMarsh.readObject(new RowReader(row, valPositions), null);
```
Comment (korlov42) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/KvMarshallerImpl.java:132]: i'm wondering why have not we done the same for `unmarshalKey`? I bet it lack of tests on putAll

File: modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerColumn.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lowka) [reply modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerColumn.java:77]: Fixed.

File: modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerReader.java (hunk 1 above)
Comment (lowka) [reply modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/MarshallerReader.java]: Done.

File: modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/Marshaller.java (hunk 2 above)
Comment (lowka) [reply modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/Marshaller.java]: Fixed.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/KvMarshallerImpl.java (hunk 5 above)
Comment (lowka) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/KvMarshallerImpl.java:148]: Fixed.

**Review by korlov42 — APPROVED** [review]

**Pull request merged**